OpenTrons/opentrons-api
# feat(app,robot-server): Subscribe to `currentlyRecoveringFrom` notifications

# Overview

Closes EXEC-458.

# Test plan

1. Push robot-server and ODD to a Flex
2. Induce a `pickUpTip` failure
3. Make sure the ODD goes into recovery mode when the `pickUpTip` failure happens
4. Make sure both recovery options work

And maybe check the app or server logs to make sure it doesn't spam network requests, especially when notifications are unavailable.

# Changelog

* Generalize the `...current_command` MQTT topic, which was "the equivalent of `GET /runs/:runId/commands?cursor=null&pageLength=1`, to `...command_links`, which is the equivalent of the `links` field of any `GET /runs/:runId/commands` query.
* Update the frontend to take advantage of that. When it sees `links` update with a new `currentlyRecoveringFrom` command ID, it will do a separate `GET /runs/{id}/commands/{id}` query to fetch the full details of the command.

# Review requests

Several frontend questions because I don't know what I'm doing there. See my GitHub comments below.

# Risk assessment

Probably at least a medium risk of breaking recent unreleased work (#15170). Although we can cover these notifications hooks with tests, it doesn't look like those tests are in a good position to make sure the UI is actually doing the correct thing overall.
